### ISSUE TITLE
Python's STDOUT isn't flushed

### ISSUE THREAD

**Issue by ikegami** [niner/inline-python-pm#32]
As [reported](https://stackoverflow.com/q/68656428/589924) on StackOverflow, Python's STDOUT isn't flushed

```perl
#!/usr/bin/env perl
use feature qw( say );

use Inline Python => <<'END_OF_PYTHON_CODE';

def test_print() -> int:
    print("In the python test print")
    return 32

END_OF_PYTHON_CODE

say "Got back ", test_print();
```

Output:

```lang-none
$ perl a.pl
In the python test print
Got back 32

$ perl a.pl | cat
Got back 32
```

Who knows what other destruction doesn't occur.

**Comment by ikegami** [comment]
One can work around this issue by adding the following to one's program:

```perl
END { Inline::Python::py_finalize(); }
```

**Comment by matthewpersico** [comment]
That doesn't quite solve the output issue. See [here](https://stackoverflow.com/a/68671281/1830614); we get all the output, but it is scrambled.

**Comment by ikegami** [comment]
> That doesn't quite solve the output issue. See [here](https://stackoverflow.com/a/68671281/1830614); we get all the output, but it is scrambled.

That's expected, completely normal, and not a bug in the module.

**Comment by matthewpersico** [comment]
> That's expected, completely normal, and not a bug in the module. 

Then it might warrant a comment in the docs. Would you take a PR for such?

**Comment by ikegami** [comment]
> > That's expected, completely normal, and not a bug in the module.
> 
> Then it might warrant a comment in the docs. Would you take a PR for such?

Whenever you have two things trying to use one resource (e.g. a terminal), coordination (e.g. synchronization) is required, or at least needs to be considered.

**Comment by matthewpersico** [comment]
> Whenever you have two things trying to use one resource (e.g. a terminal), coordination (e.g. synchronization) is required, or at least needs to be considered.

Yes, I see now. Here, we have two interpreters running, each with its own process space. Would this be an issue if we were inlining a compiled language? I would think not because then the inline'd code is just a library being executed in the Perl process space.

The distinction is important to me in terms of how I want to propose a documentation PR - I propose it would go in the Inline docs themselves under a heading like "Considering Non-Compiled Inline Modules"

**Comment by niner** [comment]
On Mittwoch, 18. August 2021 19:00:13 CEST Matthew O. Persico wrote:

> Yes, I see now. Here, we have two interpreters running, each with its own
> process space. Would this be an issue if we were inlining a compiled
> language? I would think not because then the inline'd code is just a
> library being executed in the Perl process space.

That's not correct. The Python interpreter is running the very same process as 
Perl does. In fact the Python interpreter really is just a library written in 
C. The same is actually true for Perl itself. You'll run into the same issue 
with any library that is written like it's fully in charge of a ressource like 
STDOUT.

**Comment by ikegami** [comment]
"running, each with its own process space", You mean address space? There
isn't two. There's only process and thus only one address space.

"Would this be an issue if we were inlining a compiled language?" But you
are loading a library written in a compiled language. (Not sure in which
language `python` is written. C?)
